File: src/BaekJoon/Java/baekjoon_1759.java
package BaekJoon.Java;
// https://www.acmicpc.net/problem/1759

import java.util.Arrays;
import java.util.Scanner;

public class baekjoon_1759 {
	static int L, C;
	static char[] arr;
	static boolean[] v;
	static StringBuilder sb = new StringBuilder();

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		L = sc.nextInt();
		C = sc.nextInt();
		arr = new char[C];
		v = new boolean[C];
		sc.nextLine();

		String s = sc.nextLine();
		int cnt = 0;
		for(int i = 0; i < s.length(); i++)
			if(s.charAt(i) != ' ')
				arr[cnt++] = s.charAt(i);

		Arrays.sort(arr);
		dfs("", 0);

		System.out.println(sb.toString());
		sc.close();
	}

	public static void dfs(String str, int cnt) {
		if(str.length() == L) {
			if(check(str))
				sb.append(str + "\n");
			return;
		}

		if(cnt >= arr.length)
			return;

		dfs(str + arr[cnt], cnt + 1);
		dfs(str, cnt + 1);
	}

	public static boolean check(String str) {
		int ja = 0, mo = 0;
		for(int i = 0; i < str.length(); i++) {
			char c = str.charAt(i);
			if(c == 'a' || c == 'e' || c == 'i' || c == 'u' || c == 'o')
				mo++;
			else
				ja++;
		}

		return mo > 0 && ja > 1;
	}
}

File: src/BaekJoon/Rust/baekjoon_1759.rs
use std::io::stdin;

pub fn baekjoon_1759() {
    let l_c: Vec<usize> = stdin().lines().next().unwrap().unwrap()
        .split_whitespace().map(|s| s.parse::<usize>().unwrap()).collect::<Vec<usize>>();
    let (l, c) = (l_c[0], l_c[1]);
    let mut alphabet_vector: Vec<char> = stdin().lines().next().unwrap().unwrap()
        .split_whitespace().map(|s| s.parse::<char>().unwrap()).collect::<Vec<char>>();
    
    alphabet_vector.sort_unstable();
    
    dfs(&alphabet_vector, "".to_string(), 0, l);
}

fn dfs(arr: &[char], str: String, cnt: usize, l: usize) {
    if str.len() == l {
        if check(&str) {
            println!("{}", str);
        }
        return;
    }

    if cnt >= arr.len() {
        return;
    }

    dfs(arr, format!("{}{}", str, arr[cnt]), cnt + 1, l);
    dfs(arr, str, cnt + 1, l);
}

fn check(str: &str) -> bool {
    let (mut ja, mut mo) = (0, 0);
    for c in str.chars() {
        if ['a', 'e', 'i', 'u', 'o'].contains(&c) {
            mo += 1;
        } else {
            ja += 1;
        }
    }

    mo > 0 && ja > 1
}